NoClassDefFoundError software/amazon/awssdk/services/sqs/model/ReceiveMessageResponse

Upon upgrading to 1.34. our log is filled with a NoClassDefFoundError exceptions from APM instrumentation.

Our application uses S3 via the AWS v2 SDK, but does not use or have SQS on the classpath. Attach API is used. 

Directly referencing SQS classes in the base aws client class in https://github.com/elastic/apm-agent-java/pull/2637/files#diff-c89734e3966aad23f828908ed6162f7a4077b692a3f8f5cd25a8bbd30ab0ac85 looks like an issue. Adding `software.amazon.awssdk:sqs` to the classpath works around the issue.

Here's a dump of the relevant jars from our classpath:

```
co.elastic.apm/apm-agent-api/1.34.0/c9b96292c256a406a2bc0a53cd0eedabf8a58993/apm-agent-api-1.34.0.jar
co.elastic.apm/apm-agent-attach/1.34.0/57707d3b0f0acefd3fd269db5f5e8670ae419417/apm-agent-attach-1.34.0.jar
software.amazon.awssdk/annotations/2.17.261/e61bdda35042b0a30ba6592bab3dd5cbef3de07e/annotations-2.17.261.jar
software.amazon.awssdk/apache-client/2.17.261/943d3e08e48d04b4cd0e7a7c9892b7cac0de7cbf/apache-client-2.17.261.jar
software.amazon.awssdk/arns/2.17.261/d1b00bcffde67acd4d45c277ecb6d119adddadd1/arns-2.17.261.jar
software.amazon.awssdk/auth/2.17.261/bc75acde1c19deaf5a89a7ac7cf29018a5cace0c/auth-2.17.261.jar
software.amazon.awssdk/aws-core/2.17.261/b2d22a0ca27e1fe24bc527d94445a1a34294161/aws-core-2.17.261.jar
software.amazon.awssdk/aws-json-protocol/2.17.261/b4d0c18ed2965d0e4dbef877fecbebca20156652/aws-json-protocol-2.17.261.jar
software.amazon.awssdk/aws-query-protocol/2.17.261/e702623dbe089d172d43c503f9015c8c524ee1e7/aws-query-protocol-2.17.261.jar
software.amazon.awssdk/aws-xml-protocol/2.17.261/9a679140d2569a2c7d36f5dbf8a81fc1716e7b1c/aws-xml-protocol-2.17.261.jar
software.amazon.awssdk/dynamodb-enhanced/2.17.261/eefe69a7e2e16649fb145fade5ac29e2207a273b/dynamodb-enhanced-2.17.261.jar
software.amazon.awssdk/dynamodb/2.17.261/87936410b568ba2bfdc6eb9cfa7a2d39b6f95db7/dynamodb-2.17.261.jar
software.amazon.awssdk/eventbridge/2.17.261/8e73b1547bbdf1194b1f558d9f33ae9c2202db74/eventbridge-2.17.261.jar
software.amazon.awssdk/http-client-spi/2.17.261/e548c8fcfc6b7f899d7070f084988b600954c2da/http-client-spi-2.17.261.jar
software.amazon.awssdk/json-utils/2.17.261/ad6e06cd4cf07891044694cfd9a8d5412cab59af/json-utils-2.17.261.jar
software.amazon.awssdk/metrics-spi/2.17.261/2b5d8fffb2364c6316a5607b88d4f1de8be784b6/metrics-spi-2.17.261.jar
software.amazon.awssdk/netty-nio-client/2.17.261/1922d8085bab6c8f8b924e56460de24c48847eaa/netty-nio-client-2.17.261.jar
software.amazon.awssdk/profiles/2.17.261/858a0e150ccf83201598f23f947908453ce0e537/profiles-2.17.261.jar
software.amazon.awssdk/protocol-core/2.17.261/a6cc846a3e1d082f738e81a7e64079750755463e/protocol-core-2.17.261.jar
software.amazon.awssdk/regions/2.17.261/5d60238fa52fe60f9975ec7db7613fbc57a09c24/regions-2.17.261.jar
software.amazon.awssdk/s3/2.17.261/f15c8923152e12a62ab8d9d8dc2a5dbb6f4a5599/s3-2.17.261.jar
software.amazon.awssdk/sdk-core/2.17.261/df3a59a5cb6fff0e2385ae699208636591f35c88/sdk-core-2.17.261.jar
software.amazon.awssdk/ssm/2.17.261/aeb8a323f97e34ccc0acdb417ff0b1bc0bb3afcc/ssm-2.17.261.jar
software.amazon.awssdk/sso/2.17.261/efb1cf941bd44ad3f926ecfa2e84117be6bf4791/sso-2.17.261.jar
software.amazon.awssdk/sts/2.17.261/e0dbda7de6d59ce0b7b3d0fd737813ad9b168dc5/sts-2.17.261.jar
software.amazon.awssdk/third-party-jackson-core/2.17.261/130b3b25c96ea4e058887cfa245afcf23da80e30/third-party-jackson-core-2.17.261.jar
software.amazon.awssdk/utils/2.17.261/44be242e7c2c209f86478a3bf2cd96fcb017c86f/utils-2.17.261.jar
```

```
java.lang.NoClassDefFoundError: software/amazon/awssdk/services/sqs/model/ReceiveMessageResponse
        at co.elastic.apm.agent.awssdk.v2.BaseSyncClientHandlerInstrumentation$AdviceClass.exitDoExecute(BaseSyncClientHandlerInstrumentation.java:122)
        at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.doExecute(BaseSyncClientHandler.java:167)
        at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.lambda$execute$0(BaseSyncClientHandler.java:68)
        at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.measureApiCallSuccess(BaseSyncClientHandler.java:175)
        at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.execute(BaseSyncClientHandler.java:62)
        at software.amazon.awssdk.core.client.handler.SdkSyncClientHandler.execute(SdkSyncClientHandler.java:52)
        at software.amazon.awssdk.awscore.client.handler.AwsSyncClientHandler.execute(AwsSyncClientHandler.java:63)
        at software.amazon.awssdk.services.s3.DefaultS3Client.getObject(DefaultS3Client.java:4483)
        at software.amazon.awssdk.services.s3.S3Client.getObjectAsBytes(S3Client.java:8373)
        at software.amazon.awssdk.services.s3.S3Client.getObjectAsBytes(S3Client.java:8617)
        at nz.co.sky.digital.hollow.consumer.s3.S3AnnouncementWatcher.readLatestVersion(S3AnnouncementWatcher.java:113)
        at nz.co.sky.digital.hollow.consumer.s3.S3AnnouncementWatcher.lambda$setupPollingThread$0(S3AnnouncementWatcher.java:70)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.ClassNotFoundException: software.amazon.awssdk.services.sqs.model.ReceiveMessageResponse
        at net.bytebuddy.dynamic.loading.ByteArrayClassLoader.findClass(ByteArrayClassLoader.java:403)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:587)
        at net.bytebuddy.dynamic.loading.ByteArrayClassLoader$ChildFirst.loadClass(ByteArrayClassLoader.java:1265)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
```

--- a/CHANGELOG.asciidoc
+++ b/CHANGELOG.asciidoc
@@ -38,6 +38,10 @@ endif::[]
 * Activation stack was extracted from `ElasticApmTracer` into a separate class, where it also enforces a stack depth to eliminate
 activation leaks - {pull}2783[#2783]
 
+[float]
+===== Bug fixes
+* Fix imports (leading to `NoClassDefFoundError`) in the AWS SDK instrumentation - {pull}2800[#2800]
+
 [[release-notes-1.34.0]]
 ==== 1.34.0 - 2022/09/14
 

--- a/apm-agent-plugins/apm-aws-sdk/apm-aws-sdk-2-plugin/src/main/java/co/elastic/apm/agent/awssdk/v2/BaseAsyncClientHandlerInstrumentation.java
+++ b/apm-agent-plugins/apm-aws-sdk/apm-aws-sdk-2-plugin/src/main/java/co/elastic/apm/agent/awssdk/v2/BaseAsyncClientHandlerInstrumentation.java
@@ -89,7 +89,7 @@ public class BaseAsyncClientHandlerInstrumentation extends TracerAwareInstrument
                 span.withSync(false);
             }
 
-            return new ResponseHandlerWrapper(responseHandler, sdkRequest, span);
+            return new ResponseHandlerWrapper(awsService, responseHandler, sdkRequest, span);
         }
     }
 }

--- a/apm-agent-plugins/apm-aws-sdk/apm-aws-sdk-2-plugin/src/main/java/co/elastic/apm/agent/awssdk/v2/BaseSyncClientHandlerInstrumentation.java
+++ b/apm-agent-plugins/apm-aws-sdk/apm-aws-sdk-2-plugin/src/main/java/co/elastic/apm/agent/awssdk/v2/BaseSyncClientHandlerInstrumentation.java
@@ -18,11 +18,9 @@
  */
 package co.elastic.apm.agent.awssdk.v2;
 
-import co.elastic.apm.agent.awssdk.common.IAwsSdkDataSource;
 import co.elastic.apm.agent.awssdk.v2.helper.DynamoDbHelper;
 import co.elastic.apm.agent.awssdk.v2.helper.S3Helper;
 import co.elastic.apm.agent.awssdk.v2.helper.SQSHelper;
-import co.elastic.apm.agent.awssdk.v2.helper.SdkV2DataSource;
 import co.elastic.apm.agent.awssdk.v2.helper.sqs.wrapper.MessageListWrapper;
 import co.elastic.apm.agent.bci.TracerAwareInstrumentation;
 import co.elastic.apm.agent.impl.transaction.Outcome;
@@ -33,12 +31,11 @@ import net.bytebuddy.description.type.TypeDescription;
 import net.bytebuddy.matcher.ElementMatcher;
 import software.amazon.awssdk.auth.signer.AwsSignerExecutionAttribute;
 import software.amazon.awssdk.core.SdkRequest;
+import software.amazon.awssdk.core.SdkResponse;
 import software.amazon.awssdk.core.client.config.SdkClientConfiguration;
 import software.amazon.awssdk.core.client.config.SdkClientOption;
 import software.amazon.awssdk.core.client.handler.ClientExecutionParams;
 import software.amazon.awssdk.core.http.ExecutionContext;
-import software.amazon.awssdk.services.sqs.model.ReceiveMessageRequest;
-import software.amazon.awssdk.services.sqs.model.ReceiveMessageResponse;
 
 import javax.annotation.Nullable;
 import java.net.URI;
@@ -99,9 +96,11 @@ public class BaseSyncClientHandlerInstrumentation extends TracerAwareInstrumenta
 
         @Advice.OnMethodExit(suppress = Throwable.class, inline = false, onThrowable = Throwable.class)
         public static void exitDoExecute(@Advice.Argument(value = 0) ClientExecutionParams clientExecutionParams,
+                                         @Advice.Argument(value = 1) ExecutionContext executionContext,
                                          @Nullable @Advice.Enter Object spanObj,
                                          @Nullable @Advice.Thrown Throwable thrown,
-                                         @Nullable @Advice.Return Object response) {
+                                         @Nullable @Advice.Return Object sdkResponse) {
+            String awsService = executionContext.executionAttributes().getAttribute(AwsSignerExecutionAttribute.SERVICE_NAME);
             SdkRequest sdkRequest = clientExecutionParams.getInput();
             if (spanObj instanceof Span) {
                 Span span = (Span) spanObj;
@@ -113,16 +112,15 @@ public class BaseSyncClientHandlerInstrumentation extends TracerAwareInstrumenta
                     span.withOutcome(Outcome.SUCCESS);
                 }
 
-                if (response instanceof ReceiveMessageResponse && sdkRequest instanceof ReceiveMessageRequest) {
-                    SQSHelper.getInstance().handleReceivedMessages(span, ((ReceiveMessageRequest) sdkRequest).queueUrl(), ((ReceiveMessageResponse) response).messages());
+                if ("Sqs".equalsIgnoreCase(awsService) && sdkResponse instanceof SdkResponse) {
+                    SQSHelper.getInstance().handleReceivedMessages(span, sdkRequest, (SdkResponse) sdkResponse);
                 }
+
                 span.end();
             }
 
-            if (response instanceof ReceiveMessageResponse && sdkRequest instanceof ReceiveMessageRequest) {
-                MessageListWrapper.registerWrapperListForResponse((ReceiveMessageResponse) response,
-                    SdkV2DataSource.getInstance().getFieldValue(IAwsSdkDataSource.QUEUE_NAME_FIELD, sdkRequest),
-                    SQSHelper.getInstance().getTracer());
+            if ("Sqs".equalsIgnoreCase(awsService) && sdkResponse instanceof SdkResponse) {
+                MessageListWrapper.registerWrapperListForResponse(sdkRequest, (SdkResponse) sdkResponse, SQSHelper.getInstance().getTracer());
             }
         }
     }

--- a/apm-agent-plugins/apm-aws-sdk/apm-aws-sdk-2-plugin/src/main/java/co/elastic/apm/agent/awssdk/v2/ResponseHandlerWrapper.java
+++ b/apm-agent-plugins/apm-aws-sdk/apm-aws-sdk-2-plugin/src/main/java/co/elastic/apm/agent/awssdk/v2/ResponseHandlerWrapper.java
@@ -18,19 +18,16 @@
  */
 package co.elastic.apm.agent.awssdk.v2;
 
-import co.elastic.apm.agent.awssdk.common.IAwsSdkDataSource;
 import co.elastic.apm.agent.awssdk.v2.helper.SQSHelper;
-import co.elastic.apm.agent.awssdk.v2.helper.SdkV2DataSource;
 import co.elastic.apm.agent.awssdk.v2.helper.sqs.wrapper.MessageListWrapper;
 import co.elastic.apm.agent.impl.transaction.Outcome;
 import co.elastic.apm.agent.impl.transaction.Span;
 import org.reactivestreams.Publisher;
 import software.amazon.awssdk.core.Response;
 import software.amazon.awssdk.core.SdkRequest;
+import software.amazon.awssdk.core.SdkResponse;
 import software.amazon.awssdk.core.internal.http.TransformingAsyncResponseHandler;
 import software.amazon.awssdk.http.SdkHttpResponse;
-import software.amazon.awssdk.services.sqs.model.ReceiveMessageRequest;
-import software.amazon.awssdk.services.sqs.model.ReceiveMessageResponse;
 
 import javax.annotation.Nullable;
 import java.nio.ByteBuffer;
@@ -42,8 +39,10 @@ public class ResponseHandlerWrapper<T> implements TransformingAsyncResponseHandl
     @Nullable
     private final Span span;
     private final SdkRequest sdkRequest;
+    private final String awsService;
 
-    public ResponseHandlerWrapper(TransformingAsyncResponseHandler<Response<T>> delegate, SdkRequest request, @Nullable Span span) {
+    public ResponseHandlerWrapper(String awsService, TransformingAsyncResponseHandler<Response<T>> delegate, SdkRequest request, @Nullable Span span) {
+        this.awsService = awsService;
         this.delegate = delegate;
         this.span = span;
         this.sdkRequest = request;
@@ -65,16 +64,17 @@ public class ResponseHandlerWrapper<T> implements TransformingAsyncResponseHandl
                     span.withOutcome(Outcome.SUCCESS);
                 }
 
-                if (response instanceof ReceiveMessageResponse && sdkRequest instanceof ReceiveMessageRequest) {
-                    SQSHelper.getInstance().handleReceivedMessages(span, ((ReceiveMessageRequest) sdkRequest).queueUrl(), ((ReceiveMessageResponse) response).messages());
+                if ("Sqs".equalsIgnoreCase(awsService) && response instanceof SdkResponse) {
+                    SQSHelper.getInstance().handleReceivedMessages(span, sdkRequest, (SdkResponse) response);
                 }
+
                 span.end();
             }
-            if (response instanceof ReceiveMessageResponse && sdkRequest instanceof ReceiveMessageRequest) {
-                MessageListWrapper.registerWrapperListForResponse((ReceiveMessageResponse) response,
-                    SdkV2DataSource.getInstance().getFieldValue(IAwsSdkDataSource.QUEUE_NAME_FIELD, sdkRequest),
-                    SQSHelper.getInstance().getTracer());
+
+            if ("Sqs".equalsIgnoreCase(awsService) && response instanceof SdkResponse) {
+                MessageListWrapper.registerWrapperListForResponse(sdkRequest, (SdkResponse) response, SQSHelper.getInstance().getTracer());
             }
+
         });
 
         return delegateFuture;

--- a/apm-agent-plugins/apm-aws-sdk/apm-aws-sdk-2-plugin/src/main/java/co/elastic/apm/agent/awssdk/v2/helper/SQSHelper.java
+++ b/apm-agent-plugins/apm-aws-sdk/apm-aws-sdk-2-plugin/src/main/java/co/elastic/apm/agent/awssdk/v2/helper/SQSHelper.java
@@ -25,11 +25,13 @@ import co.elastic.apm.agent.impl.transaction.Span;
 import co.elastic.apm.agent.impl.transaction.TextHeaderSetter;
 import co.elastic.apm.agent.impl.transaction.TraceContext;
 import software.amazon.awssdk.core.SdkRequest;
+import software.amazon.awssdk.core.SdkResponse;
 import software.amazon.awssdk.core.client.handler.ClientExecutionParams;
 import software.amazon.awssdk.core.http.ExecutionContext;
 import software.amazon.awssdk.services.sqs.model.Message;
 import software.amazon.awssdk.services.sqs.model.MessageAttributeValue;
 import software.amazon.awssdk.services.sqs.model.ReceiveMessageRequest;
+import software.amazon.awssdk.services.sqs.model.ReceiveMessageResponse;
 import software.amazon.awssdk.services.sqs.model.SendMessageBatchRequest;
 import software.amazon.awssdk.services.sqs.model.SendMessageBatchRequestEntry;
 import software.amazon.awssdk.services.sqs.model.SendMessageRequest;
@@ -169,6 +171,12 @@ public class SQSHelper extends AbstractSQSInstrumentationHelper<SdkRequest, Exec
             if (value != null && value.dataType().equals(ATTRIBUTE_DATA_TYPE_STRING)) {
                 consumer.accept(value.stringValue(), state);
             }
+        }
+    }
+
+    public void handleReceivedMessages(Span span, @Nullable SdkRequest sdkRequest, @Nullable SdkResponse sdkResponse) {
+        if (sdkResponse instanceof ReceiveMessageResponse && sdkRequest instanceof ReceiveMessageRequest) {
+            SQSHelper.getInstance().handleReceivedMessages(span, ((ReceiveMessageRequest) sdkRequest).queueUrl(), ((ReceiveMessageResponse) sdkResponse).messages());
         }
     }
 }

--- a/apm-agent-plugins/apm-aws-sdk/apm-aws-sdk-2-plugin/src/main/java/co/elastic/apm/agent/awssdk/v2/helper/sqs/wrapper/MessageListWrapper.java
+++ b/apm-agent-plugins/apm-aws-sdk/apm-aws-sdk-2-plugin/src/main/java/co/elastic/apm/agent/awssdk/v2/helper/sqs/wrapper/MessageListWrapper.java
@@ -19,15 +19,19 @@
 package co.elastic.apm.agent.awssdk.v2.helper.sqs.wrapper;
 
 import co.elastic.apm.agent.awssdk.common.AbstractMessageListWrapper;
-import co.elastic.apm.agent.awssdk.v2.AmazonSQSMessagingClientWrapperInstrumentation;
+import co.elastic.apm.agent.awssdk.common.IAwsSdkDataSource;
 import co.elastic.apm.agent.awssdk.v2.helper.SQSHelper;
+import co.elastic.apm.agent.awssdk.v2.helper.SdkV2DataSource;
 import co.elastic.apm.agent.configuration.MessagingConfiguration;
 import co.elastic.apm.agent.impl.ElasticApmTracer;
 import co.elastic.apm.agent.matcher.WildcardMatcher;
 import co.elastic.apm.agent.sdk.state.CallDepth;
 import co.elastic.apm.agent.sdk.weakconcurrent.WeakConcurrent;
 import co.elastic.apm.agent.sdk.weakconcurrent.WeakMap;
+import software.amazon.awssdk.core.SdkRequest;
+import software.amazon.awssdk.core.SdkResponse;
 import software.amazon.awssdk.services.sqs.model.Message;
+import software.amazon.awssdk.services.sqs.model.ReceiveMessageRequest;
 import software.amazon.awssdk.services.sqs.model.ReceiveMessageResponse;
 
 import javax.annotation.Nullable;
@@ -39,16 +43,19 @@ public class MessageListWrapper extends AbstractMessageListWrapper<Message> {
     public static final CallDepth jmsReceiveMessageCallDepth = CallDepth.get(MessageListWrapper.class);
     private static final WeakMap<ReceiveMessageResponse, List<Message>> sqsResponseToWrappedMessageListMap = WeakConcurrent.buildMap();
 
-    public static void registerWrapperListForResponse(ReceiveMessageResponse response, @Nullable String queueName, ElasticApmTracer tracer) {
-        // Wrap result only if the messages are NOT received as part of JMS.
-        if (!jmsReceiveMessageCallDepth.isNestedCallAndIncrement()) {
-            if (tracer.isRunning() && queueName != null
-                && !WildcardMatcher.isAnyMatch(tracer.getConfig(MessagingConfiguration.class).getIgnoreMessageQueues(), queueName)) {
-                sqsResponseToWrappedMessageListMap.put(response, new MessageListWrapper(response.messages(), tracer, queueName));
+    public static void registerWrapperListForResponse(@Nullable SdkRequest sdkRequest, @Nullable SdkResponse sdkResponse, ElasticApmTracer tracer) {
+        if (sdkResponse instanceof ReceiveMessageResponse && sdkRequest instanceof ReceiveMessageRequest) {
+            String queueName = SdkV2DataSource.getInstance().getFieldValue(IAwsSdkDataSource.QUEUE_NAME_FIELD, sdkRequest);
+            ReceiveMessageResponse receiveMessageResponse = (ReceiveMessageResponse) sdkResponse;
+            // Wrap result only if the messages are NOT received as part of JMS.
+            if (!jmsReceiveMessageCallDepth.isNestedCallAndIncrement()) {
+                if (tracer.isRunning() && queueName != null
+                    && !WildcardMatcher.isAnyMatch(tracer.getConfig(MessagingConfiguration.class).getIgnoreMessageQueues(), queueName)) {
+                    sqsResponseToWrappedMessageListMap.put(receiveMessageResponse, new MessageListWrapper(receiveMessageResponse.messages(), tracer, queueName));
+                }
             }
+            jmsReceiveMessageCallDepth.decrement();
         }
-        jmsReceiveMessageCallDepth.decrement();
-
     }
 
     @Nullable

